### PR TITLE
feat(c/zk): generalize halo2 prover and verifier c API

### PR DESCRIPTION
**Pull request by chokobole**
# Description

The motivation behind this PR is to simplify the naming convention for supporting `halo2` and `log_derivative_halo2` lookups. Previously, we would need to create 8 C APIs as follows:

- `tachyon_halo2_bn254_gwc_halo2_prover`
- `tachyon_halo2_bn254_gwc_log_derivative_halo2_prover`
- `tachyon_halo2_bn254_shplonk_halo2_prover`
- `tachyon_halo2_bn254_shplonk_log_derivative_halo2_prover`
- `tachyon_halo2_bn254_gwc_halo2_verifier`
- `tachyon_halo2_bn254_gwc_log_derivative_halo2_verifier`
- `tachyon_halo2_bn254_shplonk_halo2_verifier`
- `tachyon_halo2_bn254_shplonk_log_derivative_halo2_verifier`

Instead of this complicated approach, this PR unifies the following to enable switching the lookup scheme:

- `tachyon_halo2_bn254_gwc_prover` and `tachyon_halo2_bn254_shplonk_prover` to `tachyon_halo2_prover`
- `tachyon_halo2_bn254_gwc_verifier` and `tachyon_halo2_bn254_shplonk_verifier` to `tachyon_halo2_verifier`

### CODE REVIEW

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/c/zk/plonk/halo2/bn254_prover.h [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+ * @brief Creates a prover from given parameters.
+ *
+ * Initializes a prover instance for the Halo2 protocol using the specified
+ * parameters, facilitating the proof generation process for given circuits.
+ *
+ * @param pcs_type Identifier for the pcs type.
+ * * @param ls_type Identifier for the ls type.
+ * @param transcript_type The type of transcript to be used.
+ * @param k The circuit size parameter.
+ * @param params Pointer to the parameters used for prover creation.
```
Comment (ashjeong) [inline tachyon/c/zk/plonk/halo2/bn254_prover.h:70]: To align with https://github.com/kroma-network/tachyon/pull/451/commits/b6533bc69105ad41489b5c6aa344e8d6249bfdfa, could you change the lines with "Pointer to" accordingly?

**Review by ashjeong — CHANGES_REQUESTED** [review]

**Review by ashjeong — APPROVED** [review]
LGTM

File: tachyon/zk/plonk/halo2/pcs_type.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+namespace tachyon {
+namespace zk::plonk::halo2 {
+
+// THE ORDER OF ELEMENTS ARE IMPORTANT!! DO NOT CHANGE!
+// This order matches with the constants of c APIs.
+// Pleas refer to tachyon/c/zk/plonk/halo2/constants.h for details.
+enum class PCSType : uint8_t {
+  kGWC,
+  kSHPlonk,
+};
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/halo2/pcs_type.h:21]: how about giving specific number to each element?

File: tachyon/zk/plonk/halo2/ls_type.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+namespace tachyon {
+namespace zk::plonk::halo2 {
+
+// THE ORDER OF ELEMENTS ARE IMPORTANT!! DO NOT CHANGE!
+// This order matches with the constants of c APIs.
+// Pleas refer to tachyon/c/zk/plonk/halo2/constants.h for details.
+enum class LSType : uint8_t {
+  kHalo2,
+  kLogDerivativeHalo2,
+};
```
Comment (dongchangYoo) [inline tachyon/zk/plonk/halo2/ls_type.h:21]: How about giving specific value to each element?

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

File: tachyon/zk/plonk/halo2/pcs_type.h (hunk 2 above)
Comment (chokobole) [reply tachyon/zk/plonk/halo2/pcs_type.h:21]: why?

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Pull request merged**